pkg/util/push/otel.go

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	kitlog "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheusremotewrite"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/weaveworks/common/middleware"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/log"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+const (
+	pbContentType   = "application/x-protobuf"
+	jsonContentType = "application/json"
+
+	otelParseError = "otlp_parse_error"
+	maxErrMsgLen   = 1024
+
+	messageSizeLargerErrFmt = "received message larger than max (%d vs %d)"
+)
+
+func HandlerForOTLP(
+	maxRecvMsgSize int,
+	sourceIPs *middleware.SourceIPExtractor,
+	allowSkipLabelNameValidation bool,
+	push Func,
+) http.Handler {
+	return handler(maxRecvMsgSize, sourceIPs, allowSkipLabelNameValidation, push, func(ctx context.Context, r *http.Request, maxSize int, dst []byte, req *mimirpb.PreallocWriteRequest) ([]byte, error) {
+		var decoderFunc func(buf []byte) (pmetricotlp.Request, error)
+
+		logger := log.WithContext(ctx, log.Logger)
+
+		contentType := r.Header.Get("Content-Type")
+		switch contentType {
+		case pbContentType:
+			decoderFunc = func(buf []byte) (pmetricotlp.Request, error) {
+				req := pmetricotlp.NewRequest()
+				return req, req.UnmarshalProto(buf)
+			}
+
+		case jsonContentType:
+			decoderFunc = func(buf []byte) (pmetricotlp.Request, error) {
+				req := pmetricotlp.NewRequest()
+				return req, req.UnmarshalJSON(buf)
+			}
+
+		default:
+			return nil, fmt.Errorf("unsupported content type: %s, supported: [%s, %s]", contentType, jsonContentType, pbContentType)
+		}
+
+		if r.ContentLength > int64(maxSize) {
+			return nil, fmt.Errorf(messageSizeLargerErrFmt, r.ContentLength, maxSize)
+		}
+
+		reader := http.MaxBytesReader(nil, r.Body, int64(maxRecvMsgSize))
```

Comment:
is there a specific reason why this seems to use `maxSize` and `maxRecvMsgSize` interchangeably? I think it's not clear what the difference is between the two